_posts/2021-09-26-where-the-store-is-stored.md

```diff
@@ -0,0 +1,292 @@
+two years ago). 
+
+## Librairies / Framework location
+
+Another very common way of storing our store is using the locations provided by
+the main view library or framework we use (React, Vue, etc.).
+
+Let's say that we're using React here. The *go to* store location is then a
+simple `useState` hook.
+
+```js
+// app.js
+import React from 'react';
+
+import { UserContext } from './cars.js'
+
+function App({ children }) {
+  const [cars, setCars] = React.useState([]);
+
+  const addCar = car => setCars(cars => [...cars, car]);
+
+  return (
+    <MyComponent cars={cars} addCar={addCar} />
+  )
+}
+```
+
+We're not using modules to host our store anymore. React does it for us via
+`useState`.  Note that we're reactive by default here: calling `addCar` in
+`MyComponent` will rerender all `<App />`.
+
+Using `useState` alone usually doesn't scale well as we're forced to pass `cars`
+and `setCar` everywhere through props. The issue is called [props
+drilling](https://sebhastian.com/react-prop-drilling/). To overcome this, we
+can add a `Context` to make our state and its API available to every child
+components via a simple `useContext` - even if [it's not made for
+that](https://twitter.com/sebmarkbage/status/1219836431972978689). The idea is
+the same: the store is stored inside React "runtime" via `useState`.
+
+## Programatic store
+
+And finally, there's good old way: dependency injection by hand. Let's say we're
+in a full custom app (no React shenanigans) and we don't want to use modules. We
+can then craft a store instance when the app starts and pass it along everywhere
+via function parameters.
+
+Let's write a store factory very similar to the ES6 module option.
+
+```js
+// store.js
+function buildStore() {
+  const _cars = [];
+
+  return {
+    add: (car) => {
+      _cars.push(car);
+    },
+
+    list: () => {
+      return _cars;
+    },
+  }
+}
+```
+
+Now that our store's factory is ready, we just have to craft a store instance
+somewhere in our app's entrypoint and pass it along everywhere.
+
+```js
+// index.js
+
+function startApp(store) {
+  store.add('new car');
+
+  document.body.innerText = store.list().join(', ');
+
+  // ...
+}
+
+function initApp() {
+  const store = buildStore();
+
+  startApp(store);
+}
+
+initApp();
+```
+
+Of course, any function that needs to access the store will have to receive it
+via its parameters... We're right back where we left our props drilling issue.
+
+# Pros and Cons of each store location 
+
+OK so we've outlined three potential store locations. You can see me coming with
+the list of pros and cons for each of them. 
+
+Let's start with modules. Pros: it's raw ES6 (`import` / `export`) and it is
+quite simple to grasp. It's also very flexible. We can do more or less whatever
+we want in a module and every part of our codebase can use it.
+
+But there is a big constraint. Not all codebases always run in the same runtime.
```

Comment:
The `always` feels wierd to me in that part of the sentence, I would rephrase as follow:

```
The same codebase can be run in different runtimes.
```

Reply:
Ok, and the phrase is more to the point with your suggestion 👍